File: 002_GreatestCommonDivisorOfStrings.cpp
// Topic: String / Math (GCD)
// Greatest Common Divisor of Strings: Find the largest string that evenly divides two given
// strings. Time Complexity: O(n) where n = min(len1, len2) for checking, overall O(len1 + len2)
// Space Complexity: O(n) for candidate substring

#include <bits/stdc++.h>
using namespace std;

string gcdOfStrings(const string& str1, const string& str2)
{
    if (str1 + str2 != str2 + str1)
        return "";
    int a = str1.size(), b = str2.size();
    int g = std::gcd(a, b);
    return str1.substr(0, g);
}

int main()
{
    cout << gcdOfStrings("ABCABC", "ABC") << "\n";   // ABC
    cout << gcdOfStrings("ABABAB", "ABAB") << "\n";  // AB
    cout << (gcdOfStrings("LEET", "CODE").empty() ? string("") : gcdOfStrings("LEET", "CODE"))
         << "\n";  // (empty)
    return 0;
}

File: 002_GreatestCommonDivisorOfStrings.py
"""
Topic: String / Math
Greatest Common Divisor of Strings: use gcd on lengths and check concatenation property
Time Complexity: O(n + m)
Space Complexity: O(n + m)
"""

import math

def gcd_of_strings(str1: str, str2: str) -> str:
    if str1 + str2 != str2 + str1:
        return ''
    l = math.gcd(len(str1), len(str2))
    return str1[:l]


if __name__ == '__main__':
    print(gcd_of_strings('ABCABC', 'ABC'))  # 'ABC'
    print(gcd_of_strings('ABAB', 'AB'))      # 'AB'
